src/marker.rs

```diff
@@ -2,103 +2,161 @@
 //
 // based on Daniel Grunwald's https://github.com/dgrunwald/rust-cpython
 
+//! Fundamental properties of objects tied to the Python interpreter.
+//!
+//! The Python interpreter is not threadsafe. To protect the Python interpreter in multithreaded
+//! scenarios there is a global lock, the *global interpreter lock* (hereafter referred to as *GIL*)
+//! that must be held to safely interact with Python objects. This is why in PyO3 when you acquire
+//! the GIL you get a [`Python`] marker token that carries the *lifetime* of holding the GIL and all
+//! borrowed references to Python objects carry this lifetime as well. This will statically ensure
+//! that you can never use Python objects after dropping the lock - if you mess this up it will be
+//! caught at compile time and your program will fail to compile.
+//!
+//! It also supports this pattern that many extension modules employ:
+//! - Drop the GIL, so that other Python threads can acquire it and make progress themselves
+//! - Do something independently of the Python interpreter, like IO, a long running calculation or
+//! awaiting a future
+//! - Once that is done, reacquire the GIL
+//!
+//! That API is provided by [`Python::allow_threads`] and enforced via the [`Ungil`] bound on the
+//! closure and the return type. This is done by relying on the [`Send`] auto trait. `Ungil` is
+//! defined as the following:
+//!
+//! ```rust
+//! pub unsafe trait Ungil {}
+//!
+//! unsafe impl<T: Send> Ungil for T {}
+//! ```
+//!
+//! In practice this API works quite well, but it comes with some drawbacks:
+//!
+//! ## Drawbacks
+//!
+//! There is no reason to prevent `!Send` types like [`Rc`] from crossing the closure. After all,
```

Comment:
It might be nice to explain before here that the reason we have the `Send` bound is because auto traits aren't stable, and we want to implement `Ungil` for as many things as possible to make `allow_threads` nice to use. The Python Gil-bound types are all `!Send`, which is why we piggy-back off the auto trait.